src/server/pps/server/api_server.go

```diff
@@ -1243,7 +1248,70 @@ func (a *apiServer) collectDatums(ctx context.Context, job *pps.Job, cb func(*da
 }
 
 func (a *apiServer) CreateDatum(server pps.API_CreateDatumServer) (retErr error) {
-	return status.Errorf(codes.Unimplemented, "method CreateDatum not implemented")
+	msg, err := server.Recv()
+	if err != nil {
+		return errors.Wrap(err, "server receive")
+	}
+	if msg.Input == nil {
```

Comment:
It would be better to move this rule to a validation annotation on the proto message; that propagates this rule to documentation (etc.).  The validation status is checked in Recv.

Reply:
I was considering doing that. The reason I didn't is because `msg.Input` only matters for the first client request message, to compute the datums. The subsequent client messages merely indicate to the server to return the next batch of datums associated with that first message's `Input`, meaning `msg.Input` isn't necessary for every request message.

A client would have to make a separate RPC call to get datums for a different `Input`.

Reply:
I see.  This request should really be a oneof between the first request message and the subsequent request message.  input should be mandatory on the first request, and number should be mandatory on the subsequent requests:

```proto
message StartCreateDatumRequest {
    Input input = 1 [(validate.rules).message.required = true];
}
message ContinueCreateDatumRequest {
    uint32 number = 2 [(validate.rules).uint32.gt = 0];
}
message CreateDatumRequest {
    oneof body {
       StartCreateDatumRequest start = 1;
       ContinueCreateDatumRequest continue = 2;
    }
}
```

Reply:
Should the top two messages be defined inside `CreateDatumRequest`? Also, should `number` in `ContinueCreateDatumRequest` have a field tag of 1?